runners/flink/src/test/java/org/apache/beam/runners/flink/website/PipelineOptionsTableGenerator.java

```diff
@@ -0,0 +1,179 @@
+import org.apache.beam.sdk.options.Default;
+import org.apache.beam.sdk.options.Description;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.CaseFormat;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+
+/**
+ * A main class which is called by the Gradle generatePipelineOptionsTable* tasks to update the list
+ * of available pipeline options for the Beam website.
+ */
+public class PipelineOptionsTableGenerator {
+
+  private static final List<String> supportedLanguages = ImmutableList.of("java", "python");
+
+  private static class Option {
+    String name;
+    String description;
+    @Nullable String defaultValue;
+
+    public Option(String name, String description, @Nullable String defaultValue) {
+      this.name = name;
+      this.description = description;
+      this.defaultValue = defaultValue;
+    }
+  }
+
+  public static void main(String[] args) {
+    if (args.length != 1) {
+      throw new RuntimeException(
+          "Please specify the language (" + supportedLanguages + ") as the only argument.");
+    }
+    String arg = args[0].toLowerCase();
+    if (!supportedLanguages.contains(arg)) {
+      throw new RuntimeException("The language is not supported: " + arg);
+    }
+    boolean isPython = arg.equals("python");
+
+    printHeader();
+    List<Option> options = extractOptions(isPython);
+    printOptionsTable(options);
+  }
+
+  private static void printHeader() {
+    System.out.println(
+        "<!--\n"
+            + "Licensed under the Apache License, Version 2.0 (the \"License\");\n"
+            + "you may not use this file except in compliance with the License.\n"
+            + "You may obtain a copy of the License at\n"
+            + "\n"
+            + "http://www.apache.org/licenses/LICENSE-2.0\n"
+            + "\n"
+            + "Unless required by applicable law or agreed to in writing, software\n"
+            + "distributed under the License is distributed on an \"AS IS\" BASIS,\n"
+            + "WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n"
+            + "See the License for the specific language governing permissions and\n"
+            + "limitations under the License.\n"
+            + "-->");
+    System.out.println(
+        "<!--\n"
+            + "This is an auto-generated file.\n"
+            + "Use generatePipelineOptionsTableJava and generatePipelineOptionsTablePython respectively\n"
+            + "which should be called before running the tests.\n"
+            + "-->");
+  }
+
+  /**
+   * Returns the extracted list of options via reflections on FlinkPipelineOptions. Options are
+   * returned sorted in alphabetical order since Java does not guarantee any consistent order on the
+   * class methods.
+   */
+  private static List<Option> extractOptions(boolean isPython) {
+    List<Option> options = new ArrayList<>();
+    for (Method method : FlinkPipelineOptions.class.getDeclaredMethods()) {
```

Comment:
What would you say for using filter & map & collect for this? Like this: 

```
private static List<Option> extractOptions(boolean isPython) {
    return Arrays.stream(FlinkPipelineOptions.class.getDeclaredMethods())
        .filter(method -> method.getName().matches("^(get|is).*"))
        .map(method -> {
          String name = extractName(isPython, method);
          String description = extractDescription(method, name);
          String defaultValue = extractDefaultValue(method);
      return new Option(name, description, defaultValue);
    }).collect(Collectors.toList());
  }
```

and the methods: 

```
private static String extractName(boolean isPython, Method method) {
    String name = method.getName().replaceFirst("^(get|is)", "");

    if (isPython) {
      name = CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, name);
    } else {
      name = Character.toLowerCase(name.charAt(0)) + name.substring(1);
    }
    return name;
  }

  private static String extractDescription(Method method, String name) {
    Description descriptionAnnotation = method.getAnnotation(Description.class);
    if (descriptionAnnotation == null) {
      throw new RuntimeException(
          "All pipeline options should have a description. Please add one for " + name);
    }
    return descriptionAnnotation.value();
  }
```

Reply:
I prefer using iterative methods. Besides, stream is horribly slow. I appreciate the style comments! Using methods is generally more readable, but the code already uses many methods. I don't think we need to break it down further, but that's personal preference.

Reply:
My take on that is that streams let you separate the consequent steps easily and organize the code well. I don't think the speed of execution is important here. How big the class should be to feel the performance impact? 

Regarding the quantity of the methods - I don't know why is this a problem. I have to disagree one more time. If a method name explains what is done and the method does one thing only then readability increases. IMO, since developers most of the time spent working read the code rather than write it, it's worth investing time and effort to organize the code well. Notice that you wouldn't have to look inside the private methods to see what the public one does when you read the code  (thanks to good naming) so reading and getting to know it is faster.